src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,287 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using EditorBrowsableAttribute = System.ComponentModel.EditorBrowsableAttribute;
+using EditorBrowsableState = System.ComponentModel.EditorBrowsableState;
+using System.Diagnostics;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    public struct Memory<T>
+    {
+        // The highest order bit of _index is used to discern whether _arrayOrOwnedMemory is an array or an owned memory
+        // if (_index >> 31) == 1, object _arrayOrOwnedMemory is an OwnedMemory<T>
+        // else, object _arrayOrOwnedMemory is a T[]
+        private readonly object _arrayOrOwnedMemory;
+        private readonly int _index;
+        private readonly int _length;
+
+        private const int RemoveOwnedFlagBitMask = 0x7FFFFFFF;
+
+        /// <summary>
+        /// Creates a new memory over the entirety of the target array.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="array"/> is a null
+        /// reference (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArrayTypeMismatchException">Thrown when <paramref name="array"/> is covariant and array's type is not exactly T[].</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
```

Comment:
❓ Why is this restriction needed? I don't like that it means the implicit conversion operator can throw an exception that is difficult for a programmer to avoid in code.

Reply:
I am using the same restrictions as what exist for Span:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Span.cs#L4

Reply:
@jkotas Can you shed some light on this (blame shows you as introducing the original validation for `Span<T>`)?

Reply:
This check is required because Span does not support covariance. Look for covariance in Span design doc. https://github.com/dotnet/corefxlab/blob/master/docs/specs/span.md

This check has the same effect as `ref T dummy = ref array[0]` except when array has zero length.

Array covariance is rarely used, but it comes with performance penalty paid by everybody and complex implementation. We did not want to carry either of these to Span/Memory.